types/react/index.d.ts

```diff
@@ -560,6 +560,121 @@ declare namespace React {
 
     function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ComponentType<P & ClassAttributes<T>>;
 
+    //
+    // React Hooks
+    // ----------------------------------------------------------------------
+
+    // based on the code in https://github.com/facebook/react/pull/13968
+
+    // Unlike the class component setState, the updates are not allowed to be partial
+    type SetStateAction<S> = S | ((prevState: S) => S);
+    // this technically does accept a second argument, but it's already under a deprecation warning
+    // and it's not even released so probably better to not define it.
+    type Dispatch<A> = (value: A) => void;
+    // Unlike redux, the actions _can_ be anything
+    type Reducer<S, A> = (prevState: S, action: A) => S;
+    // The identity check is done with the SameValue algorithm (Object.is), which is stricter than ===
+    type IdentityCheckInputList = ReadonlyArray<any>;
+
+    // NOTE: the effect callbacks are actually allowed to return anything, but functions are treated
+    // specially. I don't think it's intended to accept async functions, should use throw instead of
+    // await to properly engage Suspense.
+    type EffectCallback = () => (void | (() => void));
+
+    interface MutableRefObject<T> {
```

Comment:
Why isn't `RefObject` the right abstraction here?

Reply:
See this https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30036#discussion_r229452449.